### PR TITLE
[WIP]OSF-Admin basic auth: User login

### PR DESCRIPTION
**Pull request by zkraime**
<h2> Purpose: </h2> 
Add basic authentication to the admin project

<b> To do 
- [x] Change back to default User (currently uses foreign key reference)
- [x] Add login/registration views
- [x] Add login/registration forms
- [x] Add authentication logic
- [x] UI 
    - [x] Get static assets working (routing issue - waiting on Lauren's webpack PR)
- [ ] Clean up code

### CODE REVIEW

File: admin/common_auth/forms.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+from django import forms
```
Comment (Ghalko) [inline admin/common_auth/forms.py]: Please use 4 spaces rather than tabs. http://cosdev.readthedocs.org/en/latest/style_guides/python.html#indentation

File: admin/base/settings/defaults.py [hunk 2]
```diff
@@ -31,10 +31,14 @@
     'django.contrib.admin',
     'django.contrib.auth',
     'django.contrib.contenttypes',
+    'django.contrib.messages',
+    'django.contrib.sessions',
     'django.contrib.staticfiles',
 
     # 3rd party
+    'admin.common_auth',
```
Comment (Ghalko) [inline admin/base/settings/defaults.py]: Does admin auth need to be in here? I don't load the spam app in https://github.com/CenterForOpenScience/osf.io/pull/4566 but it still works.

File: admin/base/settings/defaults.py [hunk 3]
```diff
@@ -56,30 +60,62 @@
...
+            ],
+        }
     }]
 
+# Database
+# https://docs.djangoproject.com/en/1.8/ref/settings/#databases
+
+DATABASES = {
+    'default': {
+        'ENGINE': 'django.db.backends.sqlite3',
```
Comment (Ghalko) [inline admin/base/settings/defaults.py]: How's the postgres progress?

File: admin/base/urls.py [hunk 4]
```diff
@@ -1,5 +1,7 @@
 from django.conf.urls import include, url, patterns
+from django.contrib import admin
 from settings import ADMIN_BASE
+admin.autodiscover()
```
Comment (Ghalko) [inline admin/base/urls.py]: What does `autodiscover` help us with? It seems that Django 1.7 and later do this automatically. https://docs.djangoproject.com/en/1.8/ref/contrib/admin/#discovery-of-admin-files

File: admin/base/urls.py [hunk 5]
```diff
@@ -10,7 +12,9 @@
     url(base_pattern,
         include(patterns('',
                          url(r'^$', views.root),
+                         url(r'^django_admin/', include(admin.site.urls)),
```
Comment (Ghalko) [inline admin/base/urls.py]: Is this a way we can obfuscate the admin login? In the future this might look like: `admin.osf.io/admin` but with potentially obfuscated names.

File: admin/common_auth/models.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+from django.db import models
+from django.contrib.auth.models import User
+
+class AdminUser(models.Model):
```
Comment (Ghalko) [inline admin/common_auth/models.py]: All users are "admin users" technically so I don't think we should make everyone `AdminUser` but rather just a `User` and the admin is implied by the site it is on. Take a look through https://docs.djangoproject.com/en/1.8/topics/auth/default/ for some information in that direction. I think we might be able to authenticate using built-in Django authentication.

File: admin/common_auth/forms.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+from django.contrib.auth import authenticate
+from django.contrib.auth.models import User
+from django.forms import ModelForm
+from models import AdminUser
+
+class RegistrationForm(ModelForm):
```
Comment (Ghalko) [inline admin/common_auth/forms.py]: Remove Registration as per discussions.

File: admin/common_auth/views.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
+from django.http import HttpResponse
+from django.contrib import messages
```
Comment (Ghalko) [inline admin/common_auth/views.py]: These two imports don't seem to be used.

File: admin/common_auth/views.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+from django.contrib.auth import authenticate, logout as logout_user, login as auth_login, views
+from django.contrib.auth.decorators import login_required, user_passes_test
+from django.contrib.auth.models import User
+from django.shortcuts import render, redirect
+
+from forms import RegistrationForm, LoginForm
+from models import AdminUser
+
+#@login_required
```
Comment (Ghalko) [inline admin/common_auth/views.py]: Why aren't we using `@login_required` here? It looks like it has a pretty flexible setup if we want to use it: https://docs.djangoproject.com/en/1.8/topics/auth/default/#the-login-required-decorator

File: manage.py [hunk 10]
```diff
@@ -1,8 +1,10 @@
 #!/usr/bin/env python
 import sys
+import os
 
 
 if __name__ == "__main__":
+    os.environ.setdefault("DJANGO_SETTINGS_MODULE", "admin.base.settings")
```
Comment (Ghalko) [inline manage.py]: This will break API, we should figure out how to run migrations without using this in the common `manage.py` file. I would suggest that we look into making an invoke script in `tasks.py` ala `adminserver` command.

**Comment by Ghalko** [comment]
Just some basic questions and requests. Looking good! :+1: 

Remember to do a couple more things before the next commit:
- `inv flake8` - travis failed because of Pep8 issues
- `[skip ci]` in your commit message - Will allow you to skip tests on travis so that it isn't doing tests on WIP PRs. That said though, it should be allowed to run before formal code reviews.
- Everyday you make changes, please commit them. I don't always remember either, but it should become reflexive.

It's all about the learning through doing. :)

**Comment by zkraime** [comment]
@Ghalko Thanks for the comments; some of these I planned to address in a commit coming in tomorrow; others I will certainly address. Appreciate the feedback, it's very helpful!

File: admin/common_auth/forms.py (hunk 1 above)
Comment (zkraime) [inline admin/common_auth/forms.py]: hmm I thought I had my settings to save tabs as four spaces; I'll make sure to check and correct

File: admin/base/settings/defaults.py (hunk 2 above)
Comment (zkraime) [inline admin/base/settings/defaults.py]: @Ghalko I needed it there for migrating, since we have AdminUser defined in models.py in common_auth.
However, as you mentioned about just using Django's built in User, we could scratch that and wouldn't have to include common_auth in installed_apps. 
The original idea behind extending the User model was to add additional fields if it were needed either now (for example, the now-deprecated idea of storing osf-user ids) or later.

File: admin/common_auth/views.py (hunk 9 above)
Comment (zkraime) [inline admin/common_auth/views.py]: Sorry didn't mean to confuse here; we will be using it

File: admin/common_auth/models.py (hunk 6 above)
Comment (zkraime) [inline admin/common_auth/models.py]: @Ghalko Please see my other comment on the original purpose of extending the User model.
However, I do agree. We only need the default User model, and using it guarantees no unexpectedness when it comes to django authentication or the admin backend.

File: admin/base/settings/defaults.py (hunk 3 above)
Comment (zkraime) [inline admin/base/settings/defaults.py]: @Ghalko Totally fine. To use postgres on local, one has to set up the db on their machine and set the proper credentials in settings.py (which ofc has to be consistent to avoid confusion). 
I was considering creating a commit with commented-out code; but I figure until this actually goes on production we don't need to worry about it and just use the default

File: admin/base/urls.py (hunk 4 above)
Comment (zkraime) [inline admin/base/urls.py]: @Ghalko You're right, we don't need to explicitly include it

File: admin/common_auth/views.py (hunk 8 above)
Comment (zkraime) [inline admin/common_auth/views.py]: HttpResponse isn't used and so should be removed,
messages is used in the template level - the relevant context will be added in a separate PR (email invitations)

File: admin/base/urls.py (hunk 5 above)
Comment (zkraime) [inline admin/base/urls.py]: @Ghalko I'm sure we could. Are we thinking of obfuscating by url name, while keeping intelligible namespaces for use in code?

File: admin/base/settings/defaults.py [hunk 11]
```diff
@@ -31,10 +31,16 @@
     'django.contrib.admin',
     'django.contrib.auth',
     'django.contrib.contenttypes',
+    'django.contrib.messages',
+    'django.contrib.sessions',
     'django.contrib.staticfiles',
+    'admin.base',
```
Comment (cwisecarver) [inline admin/base/settings/defaults.py]: `admin.base` shouldn't be an app based on the [django docs](https://docs.djangoproject.com/en/1.8/intro/tutorial01/#s-creating-a-project). I imagine it will work as one but I predict bad things in the future if it stays that way.

Comment (cwisecarver) [inline admin/base/settings/defaults.py]: Just noticed that the API does this too. :cry:

File: admin/base/settings/defaults.py (hunk 3 above)
Comment (cwisecarver) [inline admin/base/settings/defaults.py]: You could put your postgres settings in your local.py, leave the default as SQLite.

File: admin/common_auth/forms.py [hunk 12]
```diff
@@ -0,0 +1,17 @@
+from django import forms
+from django.contrib.auth import authenticate
+
+class LoginForm(forms.Form):
+    username = forms.CharField(label=(u'User Name'))
+    password = forms.CharField(label=(u'Password'), widget=forms.PasswordInput(render_value=False))
+
+    def clean(self):
```
Comment (cwisecarver) [inline admin/common_auth/forms.py]: You should be calling authenticate in the view (which you are also doing), not in the clean. You can actually remove the clean method entirely if you add `required=True` as kwargs to both of the fields. Because it inherits from `forms.Form` it will automatically call `.clean()` on the parent class.

[An example of calling authenticate and then login in the view.](https://docs.djangoproject.com/en/1.8/topics/auth/default/#django.contrib.auth.login)

[You can then use the messages framework in the view to tell them they've entered incorrect login information.](https://docs.djangoproject.com/en/1.8/ref/contrib/messages/#django.contrib.messages.add_message) and then [redirect them](https://docs.djangoproject.com/en/1.8/topics/http/shortcuts/#redirect) back to the login form.

File: admin/common_auth/urls.py [hunk 13]
```diff
@@ -0,0 +1,8 @@
+from django.conf.urls import url
+
+from . import views
+
+urlpatterns = [
+    url(r'^login/$', views.login, name='login'),
```
Comment (cwisecarver) [inline admin/common_auth/urls.py]: If you change this to `r'^login/?$'` it will not require the trailing slash, same goes for the next line. The ? makes the previous character optional.

File: admin/common_auth/views.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
+from django.contrib.auth import authenticate, logout as logout_user, login as auth_login
+from django.shortcuts import render, redirect
+
+from forms import LoginForm
+
+def login(request):
+    if request.user.is_authenticated():
+        return redirect('/admin/')
```
Comment (cwisecarver) [inline admin/common_auth/views.py]: You should use the named route for admin here. If they ever change the url of admin you don't want your name next to the blame line.

Comment (cwisecarver) [inline admin/common_auth/views.py]: Better yet, there are [login/logout views](https://docs.djangoproject.com/en/1.8/topics/auth/default/#django.contrib.auth.views.login) already built into `django.contrib.auth.views`. You could just reference those in your urls file until we need to do something they don't do.
```python
from django.contrib.auth import views as auth_views
url(r'^login/?$', auth_views.login, {'template_name': 'login.html'}),
```

File: admin/templates/base.html [hunk 15]
```diff
@@ -0,0 +1,110 @@
+{% load staticfiles %}
+<script src="{% static 'vendor/bower_components/raven-js/dist/raven.min.js' %}"></script>
```
Comment (cwisecarver) [inline admin/templates/base.html]: I commented on this in @leb2dg's PR so it's probably just a merge thing but JS should be inside the html tag. These in particular should be right before the closing head tag.

File: admin/templates/login.html [hunk 16]
```diff
@@ -0,0 +1,67 @@
+{% load render_bundle from webpack_loader %}
```
Comment (cwisecarver) [inline admin/templates/login.html]: This should extend base.html so you're not duplicating code.

File: admin/templates/login.html [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
+    <!--[if lt IE 9]>
+        <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
+        <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
+    <![endif]-->
+  </head>
+  <body class="login-page">
+    <div class="login-box">
+      <div class="login-logo">
+        <a href="../../index2.html"><b>OSF</b>Admin</a>
```
Comment (cwisecarver) [inline admin/templates/login.html]: Is there a named route for this? If so, you should use it. Hardcoded urls will come back to bite you.

File: admin/templates/login.html [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+            <div class="col-xs-4">
+              <button type="submit" class="btn btn-primary btn-block btn-flat">Sign In</button>
+            </div><!-- /.col -->
+          </div>
+        </form>
+      </div><!-- /.login-box-body -->
+    </div><!-- /.login-box -->
+
+        <!-- jQuery 2.1.4 -->
+    <script src="{% static 'js/jquery.min.js' %}"></script>
```
Comment (cwisecarver) [inline admin/templates/login.html]: If you inherit from base.html you're going to have jQuery in the page twice unless you kill this line.

File: admin/templates/base.html [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+<html>
+  <head>
+    <meta charset="UTF-8">
+    <title>OSF Admin | Dashboard</title>
+    <!-- Tell the browser to be responsive to screen width -->
+    <meta content='width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no' name='viewport'>
+    <!-- Bootstrap 3.3.4 -->
+    <link href="{% static 'css/bootstrap.min.css' %}" rel="stylesheet" type="text/css" />
+    <!-- Font Awesome Icons -->
+    <link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css" rel="stylesheet" type="text/css" />
```
Comment (cwisecarver) [inline admin/templates/base.html]: @sloria mentioned this in @leb2dg's PR as well but there's a webpack for font-awesome. This should be using that.

File: admin/base/settings/defaults.py [hunk 20]
```diff
@@ -31,10 +31,16 @@
...
+    'admin.spam',
 
     # 3rd party
+    'admin.common_auth',
```
Comment (cwisecarver) [inline admin/base/settings/defaults.py]: `admin.common_auth` isn't 3rd party, just sayin'

File: admin/base/settings/defaults.py [hunk 21]
```diff
@@ -56,30 +62,72 @@
...
+        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
+    }
+}
 
 ROOT_URLCONF = 'admin.base.urls'
 WSGI_APPLICATION = 'admin.base.wsgi.application'
 ADMIN_BASE = 'admin/'
-STATIC_URL = '{}static/'.format(ADMIN_BASE)
+STATIC_URL = '/static/'
+LOGIN_URL = '/admin/auth/login/'
```
Comment (cwisecarver) [inline admin/base/settings/defaults.py]: This doesn't appear to be used anywhere and it could be accomplished with `reverse()` to limit url hardcoding.

File: admin/base/views.py [hunk 22]
```diff
@@ -1,5 +1,10 @@
...
 
+#def root(request):
+#   return HttpResponse("Will probably need to put some front end Auth on this.")
 
-def root(request):
-    return HttpResponse("Will probably need to put some front end Auth on this.")
+@login_required
+def home(request):
+    context = {'user': request.user}
+    return render(request, 'home.html', context)
```
Comment (cwisecarver) [inline admin/base/views.py]: This could just be:
```python
return render(request, 'home.html', {'user': request.user})
```
Intermediate variables are good for clarity but only when what you're doing isn't clear to begin with.

File: admin/templates/base.html [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+    <!-- Ionicons -->
+    <link href="https://code.ionicframework.com/ionicons/2.0.1/css/ionicons.min.css" rel="stylesheet" type="text/css" />
```
Comment (cwisecarver) [inline admin/templates/base.html]: Hot linking code is dangerous. If we're going to use this we should add it to bower or download it and include it in our repo.

File: admin/templates/base.html [hunk 24]
```diff
@@ -0,0 +1,110 @@
...
+      <!-- Left side column. contains the logo and sidebar -->
+      <aside class="main-sidebar">
+
+        <!-- sidebar: style can be found in sidebar.less -->
+        <section class="sidebar">
+          <!-- Sidebar Menu -->
+          <ul class="sidebar-menu">
+            <li class="header">Menu</li>
+            <li><a href="{% url 'spam:spam' %}"><i class='fa fa-link'></i> <span>OSF Spam</span></a></li>
+            <li><a href="{}"><i class='fa fa-link'></i> <span>Pre-reg</span></a></li>
```
Comment (cwisecarver) [inline admin/templates/base.html]: Obviously this href is a placeholder but `#` is a more conventional one.

**Comment by zkraime** [comment]
@cwisecarver Thanks for the feedback, much appreciated. 
I added another subtask to clean up the code, especially the template code. A lot of it came from the adminLTE template files, but I never got around to properly formatting everything. 
I will get to that and your other comments; in the meantime I will make PRs for the rest of the code

**Comment by cwisecarver** [comment]
Thanks @zkraime. Let me know if you have any questions or if I've misunderstood anything.

File: admin/base/settings/defaults.py (hunk 21 above)
Comment (zkraime) [inline admin/base/settings/defaults.py]: @cwisecarver LOGIN_URL? 
It's used by the login_required decorator

Comment (cwisecarver) [inline admin/base/settings/defaults.py]: ah, okay. I just searched the project with silver searcher and found no reference to it. Didn't check the venv.

File: admin/templates/login.html (hunk 16 above)
Comment (zkraime) [inline admin/templates/login.html]: @cwisecarver, is there a way to do this without including the rest of base.html's page content?
